# <center>Data Understanding & Cleaning</center>

In [7]:
# Unziping the zipfile
import zipfile

with zipfile.ZipFile('bank-customers.zip', 'r') as zip_ref:
    zip_ref.extractall('data')

In [20]:
# Importing the libraries
import pandas as pd
import re

In [9]:
# Importing the dataset
df = pd.read_csv('data/Churn Modeling.csv')

In [10]:
# Checking the shape
df.shape

(10000, 14)

In [11]:
# Checking the first five rows
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [13]:
# Checking the names of the columns
df.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [17]:
# Renaming the column names
df.columns = [
    re.sub(r'(?<!^)(?=[A-Z])', '_', col).lower()
    for col in df.columns
]
df.columns

Index(['row_number', 'customer_id', 'surname', 'credit_score', 'geography',
       'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card',
       'is_active_member', 'estimated_salary', 'exited'],
      dtype='object')

In [12]:
# Checking the summary of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


### Insights
- Dataset contains 10,000 rows and 14 columns, which is sufficient for building a reliable model
- No missing values present in any column data is clean already
- There are 3 categorical features (Surname, Geography, Gender) need encoding
- There are 9 numerical features (int64) and 2 float features suitable for scaling
- Exited is the target variable and is already in binary format (0/1)

In [18]:
# Checking the statistical summary of the numerical columns
df.describe()

,row_number,customer_id,credit_score,age,tenure,balance,num_of_products,has_cr_card,is_active_member,estimated_salary,exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


### Insights
- Around 20% of customers have churned, indicating class imbalance
- credit_score shows a wide range from 350 to 850, indicating diverse customer profiles
- age varies from 18 to 92, covering a broad age group
- tenure ranges from 0 to 10 years, with an average around 5 years
- A large number of customers have zero balance, since minimum and 25th percentile are both 0
- balance has high variability due to large standard deviation, indicating possible skewness or outliers
- Most customers use 1 to 2 products, as seen from median and upper quartile values
- Majority of customers have a credit card, as the mean is around 0.70
- Customers are almost evenly split between active and inactive members
- estimated_salary has a wide spread from very low to very high values
- row_number does not carry useful information for prediction
- customer_id is unique for each customer and not useful for modeling

In [19]:
# Checking the statistical summary of categorical columns
df.describe(exclude='number')

,surname,geography,gender
count,10000,10000,10000
unique,2932,3,2
top,Smith,France,Male
freq,32,5014,5457


### Insights
- surname has 2932 unique values, indicating very high cardinality
- Most frequent surname is "Smith", but it appears only 32 times, so no strong dominance
- geography has 3 unique categories, making it a low-cardinality feature
- Majority of customers belong to France, with 5014 entries
- gender has 2 categories, making it a binary categorical feature
- Slightly more customers are Male (5457) compared to Female
- surname may not be useful for modeling due to too many unique values
- geography and gender are suitable for encoding and can be used in modeling

In [21]:
# Checking for duplicates
df.duplicated().sum()

np.int64(0)

### Insights
- No duplicate records is present

In [23]:
# Saving the dataset
df.to_csv('data/cleaned_dataset.csv', index=False)